File: src/Analyzers/VisualBasic/CodeFixes/AddExplicitCast/ArgumentFixer.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports Microsoft.CodeAnalysis.PooledObjects
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.CodeFixes.AddExplicitCast
    Partial Friend NotInheritable Class VisualBasicAddExplicitCastCodeFixProvider
        Private Class ArgumentFixer
            Inherits Fixer(Of ArgumentSyntax, ArgumentListSyntax, SyntaxNode)

            Protected Overrides Function GetExpressionOfArgument(argument As ArgumentSyntax) As ExpressionSyntax
                Return argument.GetArgumentExpression()
            End Function

            Protected Overrides Function GenerateNewArgument(oldArgument As ArgumentSyntax, conversionType As ITypeSymbol) As ArgumentSyntax
                Select Case oldArgument.Kind
                    Case SyntaxKind.SimpleArgument
                        Dim simpleArgument = DirectCast(oldArgument, SimpleArgumentSyntax)
                        Return simpleArgument.WithExpression(
                        simpleArgument.GetExpression().Cast(conversionType, Nothing))
                    Case Else
                        Return oldArgument
                End Select
            End Function

            Protected Overrides Function GenerateNewArgumentList(oldArgumentList As ArgumentListSyntax, newArguments As ArrayBuilder(Of ArgumentSyntax)) As ArgumentListSyntax
                Return oldArgumentList.WithArguments(SyntaxFactory.SeparatedList(newArguments))
            End Function

            Protected Overrides Function GetArgumentsOfArgumentList(argumentList As ArgumentListSyntax) As SeparatedSyntaxList(Of ArgumentSyntax)
                Return argumentList.Arguments
            End Function

            Protected Overrides Function GetSpeculativeSymbolInfo(semanticModel As SemanticModel, newArgumentList As ArgumentListSyntax) As SymbolInfo
                Dim parent = newArgumentList.Parent
                If TypeOf parent Is AttributeSyntax Then
                    Return semanticModel.GetSpeculativeSymbolInfo(parent.SpanStart, DirectCast(parent, AttributeSyntax))
                Else
                    Return semanticModel.GetSpeculativeSymbolInfo(parent.SpanStart, parent, SpeculativeBindingOption.BindAsExpression)
                End If
            End Function
        End Class
    End Class
End Namespace

File: src/Analyzers/CSharp/CodeFixes/AddExplicitCast/ArgumentFixer.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Extensions;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.PooledObjects;

namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.AddExplicitCast;

internal sealed partial class CSharpAddExplicitCastCodeFixProvider
{
    private class ArgumentFixer : Fixer<ArgumentSyntax, ArgumentListSyntax, SyntaxNode>
    {
        protected override ExpressionSyntax GetExpressionOfArgument(ArgumentSyntax argument)
            => argument.Expression;

        protected override ArgumentSyntax GenerateNewArgument(ArgumentSyntax oldArgument, ITypeSymbol conversionType)
            => oldArgument.WithExpression(oldArgument.Expression.Cast(conversionType));

        protected override ArgumentListSyntax GenerateNewArgumentList(ArgumentListSyntax oldArgumentList, ArrayBuilder<ArgumentSyntax> newArguments)
            => oldArgumentList.WithArguments([.. newArguments]);

        protected override SeparatedSyntaxList<ArgumentSyntax> GetArgumentsOfArgumentList(ArgumentListSyntax argumentList)
            => argumentList.Arguments;

        protected override SymbolInfo GetSpeculativeSymbolInfo(SemanticModel semanticModel, ArgumentListSyntax newArgumentList)
        {
            var newInvocation = newArgumentList.Parent!;
            return semanticModel.GetSpeculativeSymbolInfo(newInvocation.SpanStart, newInvocation, SpeculativeBindingOption.BindAsExpression);
        }
    }
}
